shim v2 doesn't pass Moby test `TestContainerStartOnDaemonRestart` (because shim v2 uses hard-coded RuncRoot)

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.
-->

**Description**

<!--
Briefly describe the problem you are having in a few paragraphs.
-->

shim v2 `io.containerd.runc.v2` doesn't pass Moby test `TestContainerStartOnDaemonRestart`: https://github.com/moby/moby/blob/88241b99893cce78a7734a19b38d468d0dcb6156/integration/container/daemon_linux_test.go#L20-L68 (https://github.com/moby/moby/issues/36145)


**Steps to reproduce the issue:**
1. Checkout moby/moby@88241b99893cce78a7734a19b38d468d0dcb6156
2. Apply the following patch to ensure that the daemon uses shim v2. This step is not needed on the cgroup v2 hosts, as cgroup v2 mode always uses shim v2.
```patch
diff --git a/daemon/daemon_unix.go b/daemon/daemon_unix.go
index 1a577276ef..1fdb924f77 100644
--- a/daemon/daemon_unix.go
+++ b/daemon/daemon_unix.go
@@ -1757,7 +1757,7 @@ func (daemon *Daemon) setupSeccompProfile() error {
 }
 
 func (daemon *Daemon) useShimV2() bool {
-       return cgroups.IsCgroup2UnifiedMode()
+       return true
 }
 
 // RawSysInfo returns *sysinfo.SysInfo .
```
3. `DOCKER_BUILD_ARGS="--build-arg CONTAINERD_COMMIT=705b8527d4945db601954246972674a25ed49347" TESTFLAGS="-test.run TestContainerStartOnDaemonRestart"  TEST_SKIP_INTEGRATION_CLI=1  make test-integration`

**Describe the results you received:**

```
--- FAIL: TestContainerStartOnDaemonRestart (2.37s)
    daemon_linux_test.go:67: assertion failed: error is not nil: Error response from daemon: OCI runtime create failed: container with id exists: 2585935e9bb8c9e5678466e4c7d001082abeacc6945d5c00cdc8f1cb713ebdd0: unknown: failed to start test container
FAIL
```

**Describe the results you expected:**

The test should succeed


**Output of `containerd --version`:**
705b8527d4945db601954246972674a25ed49347 (master as of June 16).

The issue happens with v1.3.4 as well.

**Any other relevant information:**
Moby PR https://github.com/moby/moby/pull/41115 is blocked due to this.

--- a/runtime/v2/runc/container.go
+++ b/runtime/v2/runc/container.go
@@ -20,6 +20,7 @@ package runc
 
 import (
 	"context"
+	"encoding/json"
 	"io/ioutil"
 	"os"
 	"path/filepath"
@@ -88,6 +89,10 @@ func NewContainer(ctx context.Context, platform stdio.Platform, r *task.CreateTa
 		Options:          r.Options,
 	}
 
+	if err := WriteOptions(r.Bundle, opts); err != nil {
+		return nil, err
+	}
+	// For historical reason, we write opts.BinaryName as well as the entire opts
 	if err := WriteRuntime(r.Bundle, opts.BinaryName); err != nil {
 		return nil, err
 	}
@@ -154,6 +159,39 @@ func NewContainer(ctx context.Context, platform stdio.Platform, r *task.CreateTa
 		container.cgroup = cg
 	}
 	return container, nil
+}
+
+const optionsFilename = "options.json"
+
+// ReadOptions reads the option information from the path.
+// When the file does not exist, ReadOptions returns nil without an error.
+func ReadOptions(path string) (*options.Options, error) {
+	filePath := filepath.Join(path, optionsFilename)
+	if _, err := os.Stat(filePath); err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	data, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+	var opts options.Options
+	if err := json.Unmarshal(data, &opts); err != nil {
+		return nil, err
+	}
+	return &opts, nil
+}
+
+// WriteOptions writes the options information into the path
+func WriteOptions(path string, opts options.Options) error {
+	data, err := json.Marshal(opts)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filepath.Join(path, optionsFilename), data, 0600)
 }
 
 // ReadRuntime reads the runtime information from the path

--- a/runtime/v2/runc/v1/service.go
+++ b/runtime/v2/runc/v1/service.go
@@ -208,7 +208,16 @@ func (s *service) Cleanup(ctx context.Context) (*taskAPI.DeleteResponse, error) 
 	if err != nil {
 		return nil, err
 	}
-	r := process.NewRunc(process.RuncRoot, path, ns, runtime, "", false)
+	opts, err := runc.ReadOptions(path)
+	if err != nil {
+		return nil, err
+	}
+	root := process.RuncRoot
+	if opts != nil && opts.Root != "" {
+		root = opts.Root
+	}
+
+	r := process.NewRunc(root, path, ns, runtime, "", false)
 	if err := r.Delete(ctx, s.id, &runcC.DeleteOpts{
 		Force: true,
 	}); err != nil {

--- a/runtime/v2/runc/v2/service.go
+++ b/runtime/v2/runc/v2/service.go
@@ -277,7 +277,16 @@ func (s *service) Cleanup(ctx context.Context) (*taskAPI.DeleteResponse, error) 
 	if err != nil {
 		return nil, err
 	}
-	r := process.NewRunc(process.RuncRoot, path, ns, runtime, "", false)
+	opts, err := runc.ReadOptions(path)
+	if err != nil {
+		return nil, err
+	}
+	root := process.RuncRoot
+	if opts != nil && opts.Root != "" {
+		root = opts.Root
+	}
+
+	r := process.NewRunc(root, path, ns, runtime, "", false)
 	if err := r.Delete(ctx, s.id, &runcC.DeleteOpts{
 		Force: true,
 	}); err != nil {

